Source/WebCore/platform/graphics/texmap/FloatPolygon3D.cpp

```diff
@@ -0,0 +1,119 @@
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include "config.h"
+#include "FloatPolygon3D.h"
+
+#include <wtf/TZoneMallocInlines.h>
+
+namespace WebCore {
+
+WTF_MAKE_TZONE_ALLOCATED_IMPL(FloatPolygon3D);
+
+FloatPolygon3D::FloatPolygon3D(const FloatRect& rect, const TransformationMatrix& transform)
+{
+    m_vertices.append(transform.mapPoint(FloatPoint3D(rect.minXMinYCorner())));
+    m_vertices.append(transform.mapPoint(FloatPoint3D(rect.maxXMinYCorner())));
+    m_vertices.append(transform.mapPoint(FloatPoint3D(rect.maxXMaxYCorner())));
+    m_vertices.append(transform.mapPoint(FloatPoint3D(rect.minXMaxYCorner())));
+
+    if (auto inverse = transform.inverse())
+        m_normal = inverse->transpose().mapPoint(m_normal);
+    else {
+        FloatPoint3D edge1(m_vertices[1].x() - m_vertices[0].x(), m_vertices[1].y() - m_vertices[0].y(), m_vertices[1].z() - m_vertices[0].z());
+        FloatPoint3D edge2(m_vertices[2].x() - m_vertices[0].x(), m_vertices[2].y() - m_vertices[0].y(), m_vertices[2].z() - m_vertices[0].z());
+        m_normal = edge1.cross(edge2);
+    }
+    m_normal.normalize();
+}
+
+FloatPolygon3D::FloatPolygon3D(Vector<FloatPoint3D>&& vertices, const FloatPoint3D& normal)
+    : m_vertices(WTFMove(vertices))
+    , m_normal(normal)
+{
+}
+
+// Splits the polygon into two parts relative to the given plane.
+// Algorithm:
+// - For each edge of the polygon:
+//   - Compute the signed distances of the edge's vertices from the plane.
+//   - If both vertices are on the same side of the plane, add the starting vertex to the corresponding side's list.
+//   - If the edge crosses the plane, compute the intersection point:
+//     - t = di / (di - dj)
+//     - intersectionPoint = vi + t * (vj - vi)
+//     - Add the starting vertex and the intersection point to the appropriate lists.
+// - Construct two new polygons from the collected vertices.
+std::pair<FloatPolygon3D, FloatPolygon3D> FloatPolygon3D::split(const FloatPlane3D& plane) const
+{
+    Vector<FloatPoint3D> positiveVertices;
+    Vector<FloatPoint3D> negativeVertices;
+
+    const float epsilon = std::numeric_limits<float>::epsilon(); // Tolerance for floating point comparisons
+
+    unsigned numberOfVertices = m_vertices.size();
+    for (unsigned i = 0; i < numberOfVertices; ++i) {
+        const FloatPoint3D& vi = m_vertices[i];
+        const FloatPoint3D& vj = m_vertices[(i + 1) % numberOfVertices];
+        float di = plane.distanceToPoint(vi);
+        float dj = plane.distanceToPoint(vj);
+
+        bool viPos = di > epsilon;
```

Comment:
I'd avoid those two local variables, they are only used in the following if conditions. Makes it a bit easier to read IMHO.

Reply:
Fixed